exporters/committime/collector_containerimage.py

```diff
@@ -0,0 +1,312 @@
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+#
+
+import json
+import logging
+import queue
+import subprocess
+import threading
+import time
+from typing import Dict, Iterable, Optional, Tuple
+
+from attr import define
+from openshift.dynamic.resource import ResourceField
+
+from committime import CommitMetric
+from committime.collector_base import AbstractCommitCollector
+from pelorus.timeutil import parse_guessing_timezone_DYNAMIC, to_epoch_from_string
+from provider_common.openshift import (
+    filter_pods_by_replica_uid,
+    get_and_log_namespaces,
+    get_images_from_pod,
+    get_running_pods,
+)
+
+skopeo_lock = threading.Lock()
+
+# A queue to store image URI values to be processed
+image_shas_uris_queue = queue.Queue()
+# Use to track if the image is already in the queue
+sha_in_queue = set()
+
+# Cache threshold in seconds, used by our in memory cache or storing image labels.
+# The cache may expire only when the image is not in use by running Pod anymore,
+# so we don't waste the skopeo calls to the external registries, but we still allow
+# pods to be not running for a while before we expire it's metric.
+CACHE_THRESHOLD_1_DAYS = 60 * 60 * 24
+
+# We store skopeo failures and we re-try maximum SKOPEO_MAX_RETRY times per
+# one image URI. This is to prevent too many calls to the external container
+# registries. We have a timeout here, so after some time the failed image URI
+# will be retried anyway. If the pod is not Running anymore the cache expires
+# right away.
+skopeo_failures_lock = threading.Lock()
+# The dictionary where the key is an uuid and the value a Tuple
+# where we store number of retries and the time of last check
+skopeo_failures: Dict[str, Tuple[int, float]] = {}
+SKOPEO_MAX_RETRY = 3
+CACHE_SKOPEO_FAILURE_THRESHOLD_2_DAYS = 60 * 60 * 24 * 2
+
+image_label_cache_lock = threading.Lock()
+image_label_cache: Dict[str, Tuple[Dict, float]] = {}
+
+# Store pods that are running, needed for cleanup
+running_pods_shas_lock = threading.Lock()
+running_pods_shas = set()
+
+
+class SkopeoDataException(Exception):
+    "An error that occurred Skopeo call"
+    pass
+
+
+def _add_to_cleanup_set(sha_256: str) -> None:
+    with running_pods_shas_lock:
+        running_pods_shas.add(sha_256)
+
+
+def _clear_cleanup_set() -> None:
+    with running_pods_shas_lock:
+        running_pods_shas.clear()
+
+
+def _cache_container_images_labels(sha_256: str, labels: Dict) -> None:
+    with image_label_cache_lock:
+        if sha_256 not in image_label_cache:
+            logging.debug(f"Adding SHA256 to the cache: {sha_256} ")
+            image_label_cache[sha_256] = (labels, time.time())
+
+
+def _remove_sha256_from_cache(sha_256: str) -> None:
```

Comment:
This is not used

Reply:
right, added it to have a nice interfaces, but will remove as it's not really used.